OpenRA.Mods.Common/Traits/Explodes.cs

```diff
@@ -131,30 +145,37 @@ void INotifyKilled.Killed(Actor self, AttackInfo e)
 
 		WeaponInfo ChooseWeaponForExplosion(Actor self)
 		{
-			var armaments = self.TraitsImplementing<Armament>();
-			if (!armaments.Any())
+			if (!anyArmaments)
 				return Info.WeaponInfo;
 
 			// TODO: EmptyWeapon should be removed in favour of conditions
-			var shouldExplode = !armaments.All(a => a.IsReloading);
+			var shouldExplode = false;
+
+			// PERF: Avoid LINQ
+			foreach (var a in armaments)
+			{
+				if (!a.IsReloading)
+				{
+					shouldExplode = true;
+					break;
+				}
+			}
+
 			var useFullExplosion = self.World.SharedRandom.Next(100) <= Info.LoadedChance;
 			return (shouldExplode && useFullExplosion) ? Info.WeaponInfo : Info.EmptyWeaponInfo;
```

Comment:
this can be simplified to:
```suggestion
			if (self.World.SharedRandom.Next(100) > Info.LoadedChance)
						return Info.EmptyWeaponInfo;

			// PERF: Avoid LINQ
			foreach (var a in armaments)
				if (!a.IsReloading)
					return Info.WeaponInfo;

			return Info.EmptyWeaponInfo;
```